feature/emailAuth_na

nahokyun opened the pull request:
member table에 이메일 인증을 위해 비교할 랜덤으로 생성되는 email_key와 메일 인증 여부를 체크하기위한 email_auth 칼럼을 추가하였습니다. 인증메일에서 얻어온 loginId를 통해 해당 Member 객체를 찾기 위하여 MemberService의 findByLoginId 메소드의 자료형을 Optional<Member>로 변경하였습니다. 메일 인증 완료시 member table의 mail_auth를 0에서 1로 바꾸기 위해 MemberMapper에 updateMailAuth 메소드와 관련 sql을 추가하였습니다. MemberController에는 인증메일을 보내는 emailAuth 메소드와, 전송된 메일을 통해 mail_auth를 변경하는 confirmEmail 메소드가 추가되었습니다.

랜덤으로 들어갈 난수를 생성하기위해 TempKey 클래스를 추가하였고 이메일 인증을 보낼 message 생성을 도와주는 MailHandler가 추가되었습니다.

추후 여러가지 예외를 관리하기 위해 ManageExceptionHandler를 추가하였습니다.

nahokyun committed:
feature/emailAuth_na

member table에 이메일 인증을 위해 비교할 랜덤으로 생성되는 email_key와 메일 인증 여부를 체크하기위한 email_auth 칼럼을 추가하였습니다.
인증메일에서 얻어온 loginId를 통해 해당 Member 객체를 찾기 위하여 MemberService의 findByLoginId 메소드의 자료형을 Optional<Member>로 변경하였습니다.
메일 인증 완료시 member table의 mail_auth를 0에서 1로 바꾸기 위해 MemberMapper에 updateMailAuth 메소드와 관련 sql을 추가하였습니다.
MemberController에는 인증메일을 보내는 emailAuth 메소드와, 전송된 메일을 통해 mail_auth를 변경하는 confirmEmail 메소드가 추가되었습니다.

랜덤으로 들어갈 난수를 생성하기위해 TempKey 클래스를 추가하였고 이메일 인증을 보낼 message 생성을 도와주는 MailHandler가 추가되었습니다.

추후 여러가지 예외를 관리하기 위해 ManageExceptionHandler를 추가하였습니다.

Jeonminje commented on src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -5,24 +5,40 @@
         return new ResponseEntity<>("사용가능한 아이디입니다.",HttpStatus.OK);
     }
+
+    @PostMapping("/email-auth")
+    public void emailAuth(@RequestBody Member member){
+        Optional<Member> receiver=memberService.findByLoginId(member.getLoginId());
+        if(receiver.isPresent())
+            mailService.sendEmailAuth(receiver.get());
+    }
```
파라미터로 Member 객체를 사용하신 이유가 있을까요?? 단순히 LoginId 값을 받기 위한 건지 궁금합니다!

nahokyun replied:
구현한 이메일방식의 특성상 회원가입 후 로그인직전에 이메일 인증을 통해 인증여부를 변경하는 방향이라 로그인할때와  비슷한 형태인 member형태로 받는것으로 구현하였습니다.

Jeonminje commented on src/main/java/study/movieservice/service/MailService.java:
```diff
@@ -0,0 +1,41 @@
+    public void sendEmailAuth(Member member){
+        try {
+            MailHandler sendMail = new MailHandler(javaMailSender);
+            sendMail.setSubject("[이메일 인증메일 입니다.]"); //메일제목
+            sendMail.setText(
+                    "<br>아래 [이메일 인증 확인]을 눌러주세요." +
+                            "<br><a href='http://localhost:8080/members/confirm-email?loginId="+member.getLoginId()+"&email=" + member.getEmail() +
+                            "&email_key=" + member.getEmailKey() +
+                            "' target='_blank'>이메일 인증 확인</a>");
```
MailService 클래스가 메일 제목과 내용에 의존하고 있는것같습니다

nahokyun replied:
메일을 전송하는것과 회원가입의 관심사를 분리하도록 변경하겠습니다.

sunwootest commented on src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -5,24 +5,40 @@
         return new ResponseEntity<>("사용가능한 아이디입니다.",HttpStatus.OK);
     }
+
+    @PostMapping("/email-auth")
+    public void emailAuth(@RequestBody Member member){
+        Optional<Member> receiver=memberService.findByLoginId(member.getLoginId());
+        if(receiver.isPresent())
+            mailService.sendEmailAuth(receiver.get());
+    }
```
receiver.isPresent()가 false여도 정상적으로 200 http 상태코드를 반환하네요.

nahokyun replied:
조건문을 통해 적절한 상태코드를 반환하도록 추가하겠습니다.

sunwootest commented on src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -5,24 +5,40 @@
+            mailService.sendEmailAuth(receiver.get());
+    }
+
+    @GetMapping("/confirm-email")
+    public void confirmEmail(@RequestParam String loginId,@RequestParam String email, @RequestParam("email_key") String emailKey){
```
URL이나 JSON에 snakecase를 사용하진 않습니다.

nahokyun replied:
적절한 형태로 수정하도록 하겠습니다.

sunwootest commented on src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -5,24 +5,40 @@
+    @GetMapping("/confirm-email")
+    public void confirmEmail(@RequestParam String loginId,@RequestParam String email, @RequestParam("email_key") String emailKey){
+        Optional<Member> receiver=memberService.findByLoginId(loginId);
+        if(receiver.isPresent())
+            mailService.updateMailAuth(receiver.get(),email,emailKey);
+    }
 }
```
이 부분도 마찬가지로 false일 경우 200코드를 반환합니다.

nahokyun replied:
조건문을 통해 적절한 상태코드를 반환하도록 추가하겠습니다.

sunwootest commented on src/main/java/study/movieservice/service/MailService.java:
```diff
@@ -0,0 +1,41 @@
+public class MailService {
+    private final MemberMapper memberMapper;
+    private final JavaMailSender javaMailSender;
+    @Value("${spring.mail.username}")
```
상수값이기 때문에 final로 설정하는 것이 나아보입니다. 아마 final로 설정할경우 프로퍼티값을 참조할 때 오류가 발생해서 이런 조치를 하지 않았을까 생각이 드는데 롬복을 사용안하고 빈을 주입하면 해결될겁니다.

nahokyun replied:
address또한 final로 선언후 @RequiredArgsConstructor가 아닌 생성자를 통해 해결하도록 하겠습니다.

sunwootest commented on src/main/java/study/movieservice/mail/MailHandler.java:
```diff
@@ -0,0 +1,39 @@
+package study.movieservice.mail;
+
+import org.springframework.mail.javamail.JavaMailSender;
+import org.springframework.mail.javamail.MimeMessageHelper;
+import javax.mail.MessagingException;
+import javax.mail.internet.MimeMessage;
+import java.io.UnsupportedEncodingException;
+
+public class MailHandler {
+    private final JavaMailSender mailSender;
+    private final MimeMessage message;
+    private final MimeMessageHelper messageHelper;
+
+    public MailHandler(JavaMailSender mailSender) throws MessagingException {
+        this.mailSender = mailSender;
+        message = this.mailSender.createMimeMessage();
+        messageHelper = new MimeMessageHelper(message, true, "UTF-8");
+    }
+
+    public void setSubject(String subject) throws MessagingException {
+        messageHelper.setSubject(subject);
+    }
+
+    public void setText(String htmlContent) throws MessagingException {
+        messageHelper.setText(htmlContent, true);
+    }
+
+    public void setFrom(String email, String name) throws UnsupportedEncodingException, MessagingException {
+        messageHelper.setFrom(email, name);
+    }
+
+    public void setTo(String email) throws MessagingException {
+        messageHelper.setTo(email);
+    }
+
+    public void send() {
+        mailSender.send(message);
+    }
+}
```
이 클래스는 무슨 용도인가요? 빈으로 관리되는 것도 아닌 것 같은데.

nahokyun replied:
분리의 방향이 잘못되었던것 같습니다. 삭제하고 스프링에 적합한 형태로 다시 구현해보겠습니다.

sunwootest reviewed (commented):
전반적으로 코드가독성이 떨어지는 것 같습니다. 개행도 적극적으로 사용하셔도 될 것 같습니다. 또한 이메일 인증코드는 시간이 지나면 사라져야하는데 RDB에 저장한 후 시간이 지나면 삭제하는 로직이 보이지 않습니다. 이렇게 되면 나중에 개발자의 실수로 인해 같은 번호로 계속 인증이 되어 보안에 취약할 수도 있는 로직이 될 것 같습니다.

nahokyun committed:
feature/emailAuth_na

MemberMapper에 새로운 mailKey를 업데이트하는 updateMailKey가 추가되었습니다.

nahokyun reviewed (commented):
### 추가된 내용
1. MemberMappper에 새로운 mailKey를 업데이트하는 updateMailKey가 추가되었습니다.
2. 회원가입과의 관심사 분리를 위해 MailForm이라는 클래스를 통해 메일형식을 관리하도록 추가하였습니다.
3. MemberController에서 emailAuth와 confirmEmail이 잘못된 리턴값을 받아도 200을 호출하지않도록 조건문을 추가하여 return을 하였습니다.

### 변경된 내용
1. MailService의 sendEmailAuth메소드를 sendMail로 변경하고 회원가입이라는 관심사를 분리하였습니다.
2. MemberController의 isIdDuplicated의 파라미터를 Member형식으로 변경하였습니다.
4. MemberMapper.xml의 findByLoginId쿼리 부분에서 select *의 사용을 지양하고자 필요한 칼럼들만 가져오도록 변경하였습니다.

nahokyun committed:
feature/emailAuth_na

MemberMapper의 findByLoginId에서 DB값을 받아올때 생성자 문제를 해결하기위해 @NoArgsConstructor와 @AllArgsConstructor 를 추가하였습니다.

sunwootest commented on src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -5,24 +5,57 @@
-        if(memberService.findByLoginId(loginId))
-            return new ResponseEntity<>("중복된 아이디입니다. 다시 입력해주세요", HttpStatus.CONFLICT);
-        return new ResponseEntity<>("사용가능한 아이디입니다.",HttpStatus.OK);
+    public ResponseEntity<String> isIdDuplicated(@RequestBody Member member) {
+        if (memberService.findByLoginId(member.getLoginId()).isEmpty()) {
```
아이디 중복 여부만 확인하는 api같은데 Member정보 전체를 요구할 필요가 있을까요? 아니면 다른 요구사항 때문에 그런건가요?

nahokyun replied:
@RequestBody가 String형을 원하는 형태로 받아오지 못하여 이렇게 처리하였었는데 Map<String,String>을 통해 해결할 수 있을것 같습니다.

sunwootest commented on src/main/java/study/movieservice/controller/MemberController.java:
```diff
@@ -5,24 +5,57 @@
+
+        if (receiver.isPresent()) {
+            if (mailService.updateMailAuth(receiver.get(), email, emailKey))
+                return new ResponseEntity<>("인증이 완료되었습니다.", HttpStatus.OK);
+            else
+                return new ResponseEntity<>("인증에 실패하였습니다. 메일 전송을 다시 시도해주세요", HttpStatus.BAD_REQUEST);
+        }
+        else
+            return new ResponseEntity<>("존재하지않는 아이디입니다.", HttpStatus.NOT_FOUND);
```
ResponseEntity의 내용이 너무 많아져 코드가독성이 덜어지게 되었습니다. ("인증이 완료되었습니다.", HttpStatus.OK)와 같은 내용을 enum 클래스에 static 필드로 넣어두는 것도 괜찮은 방법일 것 같네요.

nahokyun replied:
static import와 각 상황별 문구를 enum으로 선언하여 가독성을 높여보겠습니다.

sunwootest commented on src/main/java/study/movieservice/mail/MailForm.java:
```diff
@@ -0,0 +1,30 @@
+    public static String contentText(Member member,String mailSubject) {
+
+        StringBuffer buffer = new StringBuffer();
+
+        if(mailSubject.equals("회원가입인증")) {
```
회원가입인증이라는 것도 static 필드로 저장해두는것이 좋아보이네요.

sunwootest commented on src/main/java/study/movieservice/service/MemberService.java:
```diff
@@ -4,26 +4,38 @@
-        return false;
+    public Optional<Member> findByLoginId(String loginId) {
+        Member member=memberMapper.findByLoginId(loginId).get();
+        System.out.println("loginId = " + member.getLoginId());
+        System.out.println("member.getEmailKey() = " + member.getEmailKey());
+        System.out.println("member = " + member.getNickname());
```
System.out.println 사용은 지양해야합니다. 로그에 대해서 알아보세요 :)

nahokyun replied:
테스트용으로 사용하였던 println을 제거하였고 다음부턴 로그를 사용해보겠습니다.

nahokyun committed:
feature/emailAuth_na

ResponseEntity 부분의 가독성을 위해 상황별 문구를 enum으로 선언하여 간소화하고 static import를 사용하였습니다.

nahokyun commented:
### 추가된 내용
1. 상황별 텍스트의 내용을 담은 ResultEnum클래스를 추가하였습니다.
### 변경된 내용
1. 추가한 ResultEnum클래스와 static import를 통한 MemberController 클래스의 ResponseEntity 부분을 간소화하였습니다.
2. isIdDuplicated함수의 파라미터를 Map<String, String>으로 변경하여 필요한 정보만 파라미터로 받을 수 있게 변경하였습니다.
### 삭제된 내용
1. 불필요한 println문을 제거하였습니다.

sunwootest commented on src/main/java/study/movieservice/domain/ResultEnum.java:
```diff
@@ -0,0 +1,27 @@
+@Getter
+@ToString
+public enum ResultEnum {
+    SUCCESS_CAN_USE("사용가능한 아이디입니다."),
+    FAILED_CAN_USE("중복된 아이디입니다. 다시 입력해주세요"),
+
+    FAILED_AUTH("인증에 실패하였습니다. 메일 전송을 다시 시도해주세요"),
+    SUCCESS_AUTH("인증이 완료되었습니다."),
+
+    FAILED_SENDMAIL("메일전송에 실패하였습니다. 메일 전송을 다시 시도해주세요"),
+    SUCCESS_SENDMAIL("메일전송이 완료되었습니다."),
+
+    CANNOT_FOUND_ID("존재하지않는 아이디입니다.");
+
+    private final String text;
```
text와 HTTP 상태코드 정보가 같이 저장되면 더 좋을 것 같습니다 :)

nahokyun replied:
동시에 두 정보를 가지고 있도록 수정하였습니다.

sunwootest commented on src/main/java/study/movieservice/mail/MailForm.java:
```diff
@@ -0,0 +1,31 @@
+        return result;
+    }
+
+    public static String senderName() {
+        return "movieService 운영자";
+    }
```
이런건 메소드로 만들지 말고 필드로 저장하는게 나아보이네요. movieService 운영자와 같은 내용은 properties로 빼도 될 것 같고요

nahokyun replied:
application.properties에 추가하여 해당 메소드를 제거하였습니다.

sunwootest commented on src/main/java/study/movieservice/service/MailService.java:
```diff
@@ -0,0 +1,57 @@
+
+            return true;
+        }
+        else
+            return false;
```
else 문은 필요 없어보이네요

sunwootest commented on src/main/java/study/movieservice/service/MailService.java:
```diff
@@ -0,0 +1,57 @@
+
+@Service
+public class MailService {
+
+    private final MemberMapper memberMapper;
```
Mapper는 단 하나의 서비스 클래스에만 의존하게 설계하는 것이 좋습니다.

sunwootest replied:
MailService라는 클래스에 Member와 관련된 의존성이 너무 많이 보입니다 :)

nahokyun replied:
Mapper에 접근하는 메소드를 MemberService로 이동시켜 하나의 서비스 클래스에만 의존하도록 변경하였습니다.
Member와 관련된 의존성은 member의 수정가능성이 크지않을거라 생각이되고 컨트롤러 단계에서 서비스에 필요한 정보를 모두 꺼내어 String으로 넘겨주는것 또한 좋은방법은 아닐거라 생각되어 좀더 고민한뒤 추후에 변경하도록 하겠습니다.

sunwootest reviewed (commented):
다음 코드리뷰만 반영하고 approve 드릴게요. 이메일 인증이 단순한 구현은 아니어서 진행에 발목을 잡는 것 같은데 나중에 개선할 부분이 보이면 따로 이슈로 빼서 처리하면 될 것 같습니다 :)

nahokyun committed:
feature/emailAuth_na

불필요한 개행을 제거하였습니다.
Mapper에 접근하는 서비스는 하나로 설계하기위해 updateMailAuth메소드를 MemberService클래스로 이동하였습니다.
ResultEnum 클래스에 있는 기존내용에 더하여 httpstatus내용까지 부가적으로 넣어 구현하였습니다.
application.properties에 sendername 필드를 추가하여 MailForm의 senderName 메소드를 제거하였습니다.

nahokyun commented:
### 추가된 내용
1. ResultEnum 클래스에 있는 기존내용에 더하여 httpstatus내용까지 부가적으로 넣어 구현하였습니다. 
2. application.properties에 sendername 필드를 추가하였습니다. 
### 변경된 내용
1. Mapper에 접근하는 서비스는 하나로 설계하기위해 updateMailAuth메소드를 MemberService클래스로 이동하였습니다.
### 삭제된 내용
1. 불필요한 개행을 제거하였습니다.
2. MailForm의 senderName 메소드를 제거하였습니다.

sunwootest commented:
일단은 임시 approve 처리해드리겠습니다. 짝프로그래밍므로 실제 approve는 드리지 않으니 merge하지는 말아주세요.
